sklearn/partial_dependence.py

```diff
@@ -0,0 +1,577 @@
+                       grid_resolution=100, method='auto'):
+    """Partial dependence of ``target_variables``.
+
+    .. _warning_recursion_init:
+
+    .. warning::
+        The 'recursion' method only works for gradient boosting estimators,
+        and unlike the 'brute' method, it does not account for the ``init``
+        predictor of the boosting process. In practice this will produce the
+        same values as 'brute' up to a constant offset in the target
+        response, provided that ``init`` is a consant estimator (which is
+        the default). However, as soon as ``init`` is not a constant
+        estimator, the partial dependence values are incorrect.
+
+    Read more in the :ref:`User Guide <partial_dependence>`.
+
+    Parameters
+    ----------
+    est : BaseEstimator
+        A fitted classification or regression model. Multioutput-multiclass
+        classifiers are not supported.
+    target_variables : list or array-like of int
+        The target features for which the partial dependency should be
+        computed.
+    X : array-like, shape=(n_samples, n_features)
+        ``X`` is used both to generate a grid of values for the
+        ``target_variables``, and to compute the averaged predictions when
+        method is 'brute'.
+    percentiles : tuple of float, optional (default=(0.05, 0.95))
+        The lower and upper percentile used to create the extreme values
+        for the ``grid``.
+    grid_resolution : int, optional (default=100)
+        The number of equally spaced points on the grid, for each target
+        feature.
+    method : str, optional (default='auto')
+        The method used to calculate the averaged predictions:
+
+        - 'recursion' is only supported for objects inheriting from
+          `BaseGradientBoosting`, but is more efficient in terms of speed.
+          With this method, ``X`` is only used to build the
+          grid. This method does not account for the ``init`` predicor of
+          the boosting process, which may lead to incorrect values (see
+          :ref:`this warning<warning_recursion_init>`).
+
+        - 'brute' is supported for any estimator, but is more
+          computationally intensive.
+
+        - If 'auto', then 'recursion' will be used for
+          ``BaseGradientBoosting`` estimators, and 'brute' used for other
+          estimators.
+
+    Returns
+    -------
+    averaged_predictions : array, shape=(n_targets, n_points)
+        The predictions for all the points in the ``grid``, averaged over
+        all samples in X (or over the training data if ``method`` is
+        `recursion`). ``n_targets`` corresponds to the number of classes in
+        a multi-class setting, or to the number of tasks for multi-output
+        regression. For classical regression and binary classification
+        ``n_targets==1``.
+    values : seq of ndarray
+        The values with which the grid has been created. The generated grid
```

Comment:
Ideally we should say something specific like "such that `averaged_predictions[:, i]` corresponds to `values[j][i % np.prod([len(v) for v in values[j:]])]` for feature `j`". Or just reshape `averaged_predictions` so that this relationship is easier to state!